File: back-end/src/main/java/org/neuverse/service/JwtService.java
package org.neuverse.service;

import io.smallrye.jwt.build.Jwt;
import jakarta.enterprise.context.ApplicationScoped;

import java.time.Duration;
import java.util.Set;
import java.util.UUID;

@ApplicationScoped
public class JwtService {

    public String issue(UUID userId, String email, Set<String> roles, Duration ttl) {
        return Jwt.issuer("blanche-auth")
                .upn(email)
                .subject(userId.toString())
                .groups(roles == null ? Set.of("user") : roles)
                .expiresIn(ttl)
                .sign();
    }
}

File: back-end/src/main/kotlin/org/neuverse/service/JwtService.kt
package org.neuverse.service

import jakarta.enterprise.context.ApplicationScoped
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.neuverse.entity.Usuario
import io.smallrye.jwt.build.Jwt
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.temporal.ChronoUnit
import javax.crypto.spec.SecretKeySpec

@ApplicationScoped
class JwtService(

    @ConfigProperty(name = "jwt.secret")
    private val jwtSecret: String
) {

    fun generateToken(usuario: Usuario): String {
        val expiry = Instant.now().plus(1, ChronoUnit.HOURS)

        val signingKey = SecretKeySpec(
            jwtSecret.toByteArray(StandardCharsets.UTF_8),
            "HmacSHA256"
        )

        // cargo é int2 no banco -> Int no entity
        val cargoNum = usuario.cargo
        val cargoStr = cargoNum.toString() // "1", "2", "3"

        return Jwt
            .issuer("neuverse-api")
            .subject(usuario.id.toString())
            .upn(usuario.email)
            // grupos usados pelo @RolesAllowed
            .groups(setOf(cargoStr))
            // claims extras que o front usa
            .claim("role", cargoNum)
            .claim("nome", usuario.nome)
            .claim("scope", cargoNum)
            .expiresAt(expiry)
            .sign(signingKey)
    }
}
